[Jordan Quah] iP

jq1836 opened the pull request:
## DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

 - [x] Managing tasks
 - [ ] Managing deadlines (coming soon)
 - [ ] Reminders (coming soon)
 
If you Java programmer, you can use it to practice Java too. Here's the main method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

jq1836 committed:
Duke.java: Added greeting, goodbye messages, chat bot user input app loop.

jq1836 committed:
Duke.java: Moved chat bot logic out of application file.
ChatBot.java: Logic of chat bot previously from Duke.java moved here, with minor adjustments to method/variable names to better reflect what they are.

jq1836 committed:
ChatBot.java: Minor modification to echo function format, along with renaming of certain variables to reflect changes in functionality. Added list functionality as required in level-2.
ChatList.java: Created new List class customised to encapsulate the required list logic.

jq1836 committed:
Replaced ChatList.java with Task.java and TaskList.java, which better describes a list of tasks. Removed the ability to change the name of the ChatBot due to redundancy. Methodised all cases in handle methods in ChatBot.java.

jq1836 committed:
Made Task class abstract and created Deadline.java, Event.java and ToDo.java which inherits from the Task class, to support different types of tasks. Changes to ChatBot.java logic to support the new types of tasks. Removed support for the general Task.

jq1836 committed:
Abstractions made to handle the creation of tasks in a single separate method. Shortened the start method through removing redundant code. Added a method to handle unexpected user inputs.

jq1836 committed:
Included text based io redirection automated testing to test for regression in future versions of Chattus.

jq1836 committed:
Restructured project by packaging related classes together. Moved parsing functionality of ChatBot to LineParser to reduce bloat in ChatBot. Included Case enum to allow for easier labelling and handling of cases. Created InvalidCaseException and InvalidInputException to aid in handling exceptions.

jq1836 committed:
Included delete task functionality and modified test case to check for regression from this stage.

jq1836 committed:
Update all existing code to comply with Java coding standard

Current state of code does not comply entirely with Java coding
standard.

Existing code needs to be modified to meet standards.

Existing code has been modified to meet standards.

jq1836 committed:
Remove unnecessary use of 'this' keyword

Current state of code has many unnecessary usage of 'this' keyword.

Removing these unnecessary usage of 'this' keyword reduces
unnecessary noise in the code.

Remove unnecessary use of 'this' keyword.

jq1836 committed:
Update import statements to follow Java conventions

There were 2 import statement that was out of lexicographical order.

Update import statements to follow lexicographical order.

jq1836 committed:
Fix text based automated testing

Restructuring of the project through packaging of relevant files
has resulted in the path in the runtest.sh file to be inconsistent
with the changes made.

Modify the path in the runtest.sh file to be consistent with
changes made.

jq1836 committed:
Format switch case segments to follow course's Java conventions

Switch case statement does not follow course's Java conventions.

Format switch case segments to satisfy above subject.

jq1836 committed:
Implement file save and load functionality

Application currently does not have any functionality to save or
load files.

Adding save and load functionality will allow for users to retain
any tasks that they input.

Lets.
1. Implement taskfilemanager package with classes which serve to
enable file save and load functionality.
2. Rename LineParser.java to CommandParser.java to avoid confusion
between TaskLineParser.java and LineParser.java.
3. Modify existing code to support file save and load functionality.
4. Add new exception TaskFileException.

jq1836 committed:
Update gitignore to ignore duke.txt file

With the recent changes, duke.txt will be generated when file is
saved. It is unnecessary for git to track the file.

Add duke.txt to gitignore.

jq1836 committed:
Update ChatBot.java to adhere to project's Java conventions

ChatBot.java was modified in a previous commit, which resulted in
non-adherence to the project's Java conventions.

Swap positions of fields in ChatBot.java.

jq1836 committed:
Implement rigid date and time functionality

Event.java and Deadline.java currently uses strings to store
date and time, which is of little use.

Making use of proper date and time abstractions would enhance the
usefulness of the product.

Lets,
1. Use java.time.LocalDateTime instead of String representation of
date and deadline representations in Event.java and Deadline.java.
2. Modify LineParser.java to support the change in representations.
3. Implement check command which displays all tasks that are on the
specified date.

jq1836 committed:
Merge branch 'branch-Level-7'

jq1836 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/task/Deadline.java
#	src/main/java/task/Event.java

Resolution:
1. Modified format slightly to allow for both branches to be
compatible.
2. Chose the appropriate constructors within Deadline.java and
Event.java.

jq1836 committed:
Fix gitignore

In a recent commit, the gitignore was modified to include
/data/duke.txt in order to stop tracking files generated by the
application.

Change gitignore to include /data instead of /data/duke.txt.

jq1836 committed:
Create Command abstraction and reorganize packages

Currently, the various commands that users input will be tagged by
enum Case.java.

There is a better way of abstracting the commands that users input.

Lets,
1. Add abstract Command.java and all subclasses of it.
2. Modify all existing parsing and code to support the new
abstraction of commands.
3. Remove all deprecated classes.
4. Reorganize packages to fit new abstraction.

jq1836 committed:
Consolidate parser classes into single parser class

Currently, parsing is being handled by TimedTask.class,
TaskLineParser.java and CommandParser.java.

Consolidating all parsing jobs to a single class results in
easier debugging and better segmentation of code.

Lets,
1. Combine all parsing functionality into a single parser class.
2. Prepare for implementation of UI.java by creating the class and
moving TaskList and Storage into the new main application class,
Duke.java.

jq1836 committed:
Implement Ui class and update Duke.java

Duke.java class currently does not support Ui class.

Lets,
1. Implement Ui.java class.
2. Update Duke.java class to support Ui.java class.
3. Minor reformatting of parser to remove whitespace in front
of task descriptions.

jq1836 committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

jq1836 committed:
Add JUnit tests for Event.java and ToDo.java

Currently, there are no unit tests to detect for any undesired behavior.

JUnit tests will help detect future bugs from change in code.

Add JUnit tests for Event.java and ToDo.java.

jq1836 committed:
Update build.gradle to include shadowJar

Currently, the chat bot is really for a v0.1 release.

Modify build.gradle to include shadowJar.

jq1836 committed:
Implement find functionality

Currently, there is no search substring functionality for the chat bot.

Adding a find command would improve the usability of the chat bot.

Lets,
1. Implement FindCommand.java
2. Modify Task and TaskList to support finding substring
3. Update parser to recognise find command

jq1836 committed:
Update code to be consistent with Java coding standards

Code was not consistent with Java coding standards.

Update code to be consistent with Java coding standards.

jq1836 committed:
Write out Java Docs for all existing classes

Currently, the code does not have any Java Docs.

Write Java Docs for all classes.

jq1836 committed:
Merge branch 'branch-A-JavaDoc'

jq1836 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/chatbot/Duke.java
#	src/main/java/duke/chatbot/command/AddDeadlineCommand.java
#	src/main/java/duke/chatbot/command/AddEventCommand.java
#	src/main/java/duke/chatbot/command/AddToDoCommand.java
#	src/main/java/duke/chatbot/command/CheckDateCommand.java
#	src/main/java/duke/chatbot/command/DeleteCommand.java
#	src/main/java/duke/chatbot/command/ExitCommand.java
#	src/main/java/duke/chatbot/command/InvalidInputCommand.java
#	src/main/java/duke/chatbot/command/ListCommand.java
#	src/main/java/duke/chatbot/command/MarkCommand.java
#	src/main/java/duke/chatbot/command/UnmarkCommand.java
#	src/main/java/duke/chatbot/common/Message.java
#	src/main/java/duke/chatbot/data/task/TimedTask.java
#	src/main/java/duke/chatbot/parser/Parser.java
#	src/main/java/duke/chatbot/storage/TaskFileSaver.java
#	src/main/java/duke/chatbot/ui/Ui.java

jq1836 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/chatbot/Duke.java
#	src/main/java/duke/chatbot/data/task/TaskList.java

nealetham commented on src/main/java/duke/chatbot/common/Message.java:
```diff
@@ -0,0 +1,26 @@
+/**
+ * Contains the messages that are used by the application.
+ */
+public class Message {
```
I like how you abstracted all the non object-specific messages into this file 👍

nealetham commented on src/main/java/duke/chatbot/data/task/Task.java:
```diff
@@ -0,0 +1,73 @@
+    /**
+     * Sets the task to the done state.
+     */
+    public void mark() {
```
Perhaps you could use a more intuitive method name such as setDone()?

nealetham commented on src/main/java/duke/chatbot/data/task/Task.java:
```diff
@@ -0,0 +1,73 @@
+    /**
+     * Sets the task to the not done state.
+     */
+    public void unmark() {
```
As per the comment on the mark() method.

nealetham commented on src/test/java/duke/chatbot/data/task/EventTest.java:
```diff
@@ -0,0 +1,80 @@
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class EventTest {
```
Very comprehensive tests 👍 
Maybe consider adding javadocs to describe them?

nealetham reviewed (commented):
Overall, I found your code easy to read for the most part except a few places
where the nesting was too deep (Parser.java). Generally, there aren't many coding standard violations too. Great effort in abstracting so many classes. 👍

jq1836 committed:
Add Java Docs to FindCommand class

FindCommand class currently does not include Java Docs.

Add Java Docs to FindCommand class.

jq1836 committed:
Add new message for FindCommand

Currently, a successful FindCommand will result in the list of
tasks matching the query being printed, without any header message.
Additionally, FindCommand and CheckDateCommand do not currently
have a message for when no tasks match the query.

Adding the FindCommand custom message and the message for when
no tasks match the query will enhance user experience.

Lets,
1. Add custom message MESSAGE_FIND_KEYWORD.
2. Add logic for CheckDateCommand and FindCommand that supports
no task matching query message, MESSAGE_EMPTY_LIST.

jq1836 committed:
Add check style to project and update code to comply with it

Currently, the project lacks a check style.

Lets,
1. Add check style to project.
2. Fix any existing code style violation that is detected by the
check style.

jq1836 committed:
Merge branch 'branch-A-CheckStyle'

jq1836 committed:
Install JavaFX into the project

Currently, project is using text based UI.

Switching over to JavaFX based UI will enhance user experience.

Install JavaFX.

jq1836 committed:
Update Duke.java to use Java Varargs

Currently, the project does not make use of Java Varargs.

Update Duke.java to include Java Varargs.

jq1836 committed:
Merge branch 'branch-A-Varargs'

jq1836 committed:
Update Java Docs to include @author tag

Currently, Java Docs in the project does not include @author tag.

Add @author tag to Java Docs.

jq1836 committed:
Update names of mark and unmark methods in Task.java

The current mark and unmark method names may not be clear.

Lets,
1. Change mark method name to markDone
2. Change unmark method name to markUndone

jq1836 committed:
Add GUI to the project

Currently, the project does not have a GUI.

Lets,
1. Create a GUI.
2. Create MessageBuilder class which is a StringBuilder which
supports new lines.
3. Update all existing Command classes to make use of the
MessageBuilder class.
4. Change Java Docs @author tag to jq1836.
5. Modify existing code to support new GUI.

jq1836 committed:
Merge branch 'branch-Level-10'

ReubenChay commented on src/main/java/duke/chatbot/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,38 @@
+package duke.chatbot.command;
```
really like this interesting way of managing the inputs with command class!

jsincorporated replied:
Agreed with the above comment, really good use of OOP 👍

ReubenChay commented on src/main/java/duke/chatbot/data/exception/InvalidStorageFileException.java:
```diff
@@ -0,0 +1,11 @@
+ * not in the proper format.
+ * @author jq1836
+ */
+public class InvalidStorageFileException extends FileNotFoundException {
```
correct me if im wrong but wouldn't it be fine to just throw a filenotfound whenever a wrong file format is provided?

ReubenChay commented on src/main/java/duke/chatbot/util/Parser.java:
```diff
@@ -0,0 +1,159 @@
+import duke.chatbot.data.task.Event;
+import duke.chatbot.data.task.TaskList;
+import duke.chatbot.data.task.ToDo;
+
```
really love the use of OOP to abstract everything but maybe these 30 lines of import statements might indicate over-abstraction? not too big of an issue either way

ReubenChay reviewed (commented):
extensive use of OOP and really thorough testing. LGTM!

jsincorporated commented on src/main/java/duke/chatbot/data/task/Task.java:
```diff
@@ -0,0 +1,73 @@
+     * Returns a string that is written onto a text file to be stored.
+     * @return A string that is written onto a text file to be stored.
+     */
+    public String encode() {
```
I like how you used the name to explain what the method does

jsincorporated reviewed (commented):
Overall code is clean and follows coding standard! LGTM :)

jq1836 committed:
Update GUI and prepare for v0.2 release

1. Update build.gradle file's main .java file to launcher.java.
2. Swapped the images of Duke and User in the GUI.

jq1836 committed:
Add unit testing for Deadline and update for Event and ToDo

Currently, there is no unit testing for Deadline class.
Additionally, all these classes do not have the desired equals
method and unit tests for the method.

Adding unit tests for these classes will allow for detection of
regression for these classes in the future.

Lets,
1. Add unit tests for Deadline.java
2. Update unit tests for Event.java
3. Update unit tests for ToDo.java
4. Add equals method to Deadline.java
5. Add equals method to Event.java
6. Add equals method to ToDo.java

jq1836 committed:
Refactor code and add links in Javadoc

Currently, the Javadoc comments have different maximum line
lengths. Javadoc comments also do not have links.

Standardizing Javadoc comments maximum line length will make the
code cleaner and more readable. Adding links to the Javadoc
comments will allow for easier navigation between classes.

Lets,
1. Refactor Javadoc comments
2. Add links in Javadoc comments

jq1836 committed:
Update Command and Parser classes to make use of Java regex

Currently, Parser, Command and its subclasses do not make use of
Java regex.

Making use of Java regex will allow for ease of creating blueprints
for future commands.

Lets,
1. Update Parser.java to make use of Java regex
2. Update Command and its subclasses to make use of Java regex
3. Add new messages to Message.java
4. Update InvalidInputCommand.java to be able to have its own
error message
5. Update InvalidInputException.java to support point 4
6. Add some unit tests for Parser.java
7. Create AddTaskCommand.java which encapsulates a command that
adds tasks
8. Update Duke.java to make use of the new InvalidInputException
and InvalidInputCommand
9. Update MessageBuilder.java with two new methods to include the
ability to add single lines one at a time
10. Update TaskList.java to make use of updated MessageBuilder.java

jq1836 committed:
Implement FilterCommand.java

Currently, the filter functionality in the application is hard
coded.

By implementing it using Java util Predicates, we will be able to
make it easier for different types of filter functionality to be
implemented in the future.

Lets,
1. Implement FilterCommand.java
2. Modify existing code to support point 1

jq1836 committed:
Add Java assertions

Currently, the code does not make use of Java assertions.

Including Java assertions will allow for bugs to be detected with
greater ease.

Add Java assertions.

jq1836 committed:
Merge pull request #8 from jq1836/branch-A-Assertions

Add Java assertions

jq1836 committed:
Improve code quality

Currently, the code is cluttered in some segments.

Refactoring the code allows for greater readability of the code.

Improve code quality with SLAP and abstraction.

jq1836 committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/chatbot/command/AddTaskCommand.java
#	src/main/java/duke/chatbot/command/DeleteCommand.java
#	src/main/java/duke/chatbot/command/FilterCommands/FilterCommand.java
#	src/main/java/duke/chatbot/command/ListCommand.java
#	src/main/java/duke/chatbot/command/MarkCommand.java
#	src/main/java/duke/chatbot/command/UnmarkCommand.java

jq1836 committed:
Merge pull request #9 from jq1836/branch-A-CodeQuality

Improve code quality

jq1836 committed:
Create gradle.yml

Currently, the project does not have continuous integration.

Adding continuous integration will allow for gradle tests to check if commit or pull request is suitable for merging with master branch.

Create gradle.yml

jq1836 committed:
Remove unused import in DeleteCommand.java

Currently, in DeleteCommand.java, there is an unused import.

Remove unused import.

jq1836 committed:
Fix regression

Currently, a recent commit caused regression.

Fix regression.

jq1836 committed:
Update ParserTest.java to comply with checkstyle

Currently, due to the scope of the checkstyle not being set to all
Java code in the project, the code in the unit test may violate the
checkstyle.

Update code in ParserTest.java to comply with checkstyle.

jq1836 committed:
Merge pull request #10 from jq1836/branch-A-CI

Create gradle.yml

jq1836 committed:
Add data file switch, create and delete functionality

Currently, Duke makes use of only duke.txt.

Adding the ability to create, delete and change to different data
files allows a greater deal of flexibility for users.

Implement changes to support multi-file application.

jq1836 committed:
Fix style issues before merge into master branch

Currently, there are issues regarding style in the code.

Fix style issues in code.

jq1836 committed:
Fix style issue

Missed out 1 style issue in the previous commit.

Fix style issue.

jq1836 committed:
Merge pull request #11 from jq1836/branch-C-FlexibleDataSource

Add data file switch, create and delete functionality

jq1836 committed:
Remove text-ui-test directory

Currently, the application does not make use of text-based ui
anymore.

Removing the directory removes clutter from the project.

Remove text-ui-test directory.

jq1836 committed:
Update code with code quality feedback

Received code quality feedback from soc-se-bot-blue.

Update code using feedback.

jq1836 committed:
Change user image

Current user image might be copyright material.

Replace user image.

jq1836 committed:
Rename Duke to Rattus

Currently, the chatbot's name is Duke.

To add personality to the chatbot, change the name of the chatbot.

Rename Duke to Rattus.

jq1836 committed:
Rename Duke to Rattus in GUI code

Currently, there are parts of the GUI code with Duke.

Rename Duke to Rattus in GUI code.

jq1836 committed:
Update README.md

Currently, README.md has not been updated since the start of the
project.

Update README.md.

jq1836 committed:
Update import statement order

Due to major refactoring in earlier commit, import statements
went out of order.

Update import statement order.

jq1836 committed:
Fix broken link in README.md

There is a broken link in README.md.

Fix broken link in README.md.
